src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,683 @@
+    // as possible.  This will also perform opportunistic shielding if the
+    // transaction strategy permits.
+
+    CAmount maxSaplingAvailable = spendableMut.GetSaplingTotal();
+    CAmount maxOrchardAvailable = spendableMut.GetOrchardTotal();
+    uint32_t orchardOutputs{0};
+
+    // we can only select Orchard addresses if there are sufficient non-Sprout
+    // funds to cover the total payments + fee.
+    bool canResolveOrchard =
+        params.GetConsensus().NetworkUpgradeActive(anchorHeight, Consensus::UPGRADE_NU5)
+        && spendableMut.Total() - spendableMut.GetSproutTotal() >= targetAmount;
+    std::vector<ResolvedPayment> resolvedPayments;
+    std::optional<AddressResolutionError> resolutionError;
+    for (const auto& payment : payments) {
+        std::visit(match {
+            [&](const CKeyID& p2pkh) {
+                if (strategy.AllowRevealedRecipients()) {
+                    resolvedPayments.emplace_back(
+                            std::nullopt, p2pkh, payment.GetAmount(), payment.GetMemo(), false);
+                } else {
+                    resolutionError = AddressResolutionError::TransparentRecipientNotAllowed;
+                }
+            },
+            [&](const CScriptID& p2sh) {
+                if (strategy.AllowRevealedRecipients()) {
+                    resolvedPayments.emplace_back(
+                            std::nullopt, p2sh, payment.GetAmount(), payment.GetMemo(), false);
+                } else {
+                    resolutionError = AddressResolutionError::TransparentRecipientNotAllowed;
+                }
+            },
+            [&](const SproutPaymentAddress&) {
+                resolutionError = AddressResolutionError::SproutRecipientsNotSupported;
+            },
+            [&](const SaplingPaymentAddress& addr) {
+                if (strategy.AllowRevealedAmounts() || payment.GetAmount() < maxSaplingAvailable) {
+                    resolvedPayments.emplace_back(
+                            std::nullopt, addr, payment.GetAmount(), payment.GetMemo(), false);
+                    if (!strategy.AllowRevealedAmounts()) {
+                        maxSaplingAvailable -= payment.GetAmount();
+                    }
+                } else {
+                    resolutionError = AddressResolutionError::RevealingSaplingAmountNotAllowed;
+                }
+            },
+            [&](const UnifiedAddress& ua) {
+                if (canResolveOrchard
+                    && ua.GetOrchardReceiver().has_value()
+                    && (strategy.AllowRevealedAmounts() || payment.GetAmount() < maxOrchardAvailable)
+                    ) {
+                    resolvedPayments.emplace_back(
+                        ua, ua.GetOrchardReceiver().value(), payment.GetAmount(), payment.GetMemo(), false);
+                    if (!strategy.AllowRevealedAmounts()) {
+                        maxOrchardAvailable -= payment.GetAmount();
+                    }
```

Comment:
This is confusing to me. If both `strategy.AllowRevealedAmounts()` and `payment.GetAmount() <= maxOrchardAvailable`, what do we do?

Reply:
See https://github.com/zcash/zcash/pull/6377/files#r1142654730